Add support for regions to the segment builder.

glennw opened the pull request:
This allows specifying an inner rect for the segment builder,
which defines the main primitive as a 9-patch. Then, when
segments are created, the region they belong to is passed to
the segment closure.

This allows us to do things like build segments for a box-shadow
primitive, where the box-shadow is made up of sub-regions (i.e.
the corners, edges and inner region are separate regions).

In the case of a fast box shadow path, the region can be used
to identify the UV sub-rect from the blurred image to draw with.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/webrender/2306)
<!-- Reviewable:end -->

glennw commented:
r? @kvark

This contains the logic and unit tests for using regions in the segment builder.

The box-shadow code to make use of this isn't quite ready. I thought I'd open this PR up now
since it might be easier to review as an isolated patch (most of the new lines are tests). If
you'd prefer to wait until the box-shadow patch that uses this functionality is ready, it's fine
to just leave this for now.

mstange commented:
> This allows specifying an inner rect for the segment builder, which defines the main primitive as a 9-patch.

I haven't read the code, but based on this sentence, I don't think this is sufficient. If you have a large top left corner and a small top right corner, then doesn't this mean that the top right corner's segment size will be unnecessarily large? And if the top left corner and bottom right corner overlap, there's not even a valid inner rect that describes the necessary segmentation.

glennw commented:
@mstange You're correct - this initial implementation uses a nine-patch, since that will handle the common case (and won't segment at all in the cases you described above). I've done it this way to keep the initial implementation simple, but the segment builder internally can handle arbitrary region definitions, so we can expand it later. Does that sound OK?

mstange commented:
I see, that makes sense. I just didn't want us to get stuck after going down the wrong path. But it sounds like this is actually just a stop on the right path.

kvark commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[webrender/src/segment.rs, line 43 at r1](https://reviewable.io:443/reviews/servo/webrender/2306#-L2waXiSSS2oUr19Kok5:-L2waXiSSS2oUr19Kok6:bu2oxqp) ([raw file](https://github.com/servo/webrender/blob/b2a91765fd52cc16b4ff79334379b46bc40d4484/webrender/src/segment.rs#L43)):*
> ```Rust
>     Begin,
>     End,
>     BeginRegion,
> ```

could we have some internal docs here explaining the difference between `Begin` and `BeginRegion`?

---

*[webrender/src/segment.rs, line 53 at r1](https://reviewable.io:443/reviews/servo/webrender/2306#-L2waN_onuAVhdKhPlXp:-L2waN_pe-2g8R3dxgK5:b-79slec) ([raw file](https://github.com/servo/webrender/blob/b2a91765fd52cc16b4ff79334379b46bc40d4484/webrender/src/segment.rs#L53)):*
> ```Rust
>     fn cmp(&self, other: &EventKind) -> cmp::Ordering {
>         match (*self, *other) {
>             (EventKind::BeginRegion, EventKind::BeginRegion) => {
> ```

I recall that the previous attempt to get this derived automatically didn't succeed, but we haven't figured out why.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/webrender/2306#-:-L2wbloJ-NpCendJ5WvH:bnfp4nl)*
<!-- Sent from Reviewable.io -->

glennw commented:
Review status: 1 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[webrender/src/segment.rs, line 43 at r1](https://reviewable.io:443/reviews/servo/webrender/2306#-L2waXiSSS2oUr19Kok5:-L2xJugd48cgLPVyRaFr:b-3ya5bf) ([raw file](https://github.com/servo/webrender/blob/b2a91765fd52cc16b4ff79334379b46bc40d4484/webrender/src/segment.rs#L43)):*
<details><summary><i>Previously, kvark (Dzmitry Malyshau) wrote…</i></summary><blockquote>

could we have some internal docs here explaining the difference between `Begin` and `BeginRegion`?
</blockquote></details>

I added a comment and renamed Begin to BeginClip so it's a bit clearer what they represent.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/webrender/2306)*
<!-- Sent from Reviewable.io -->

glennw commented:
Thanks, added a comment and renamed the enum variants to be a bit more descriptive.

kvark commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 7fef373 has been approved by `kvark`

<!-- @bors-servo r=kvark 7fef3730780b4a5612cf98e16c0f73c209b5ebbe -->

bors-servo commented:
:hourglass: Testing commit 7fef3730780b4a5612cf98e16c0f73c209b5ebbe with merge ef420106c1d28a7f1b342201b837d4b5f1a53831...

bors-servo commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/servo/webrender/build/1.0.7496), [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/VKNmxer2Q8Ki_gYGXbMrGQ)
Approved by: kvark
Pushing ef420106c1d28a7f1b342201b837d4b5f1a53831 to master...
